### DISCUSSION TITLE
Uploads should have the same permission control as content types

### DISCUSSION THREAD

**Discussion by aahung** [strapi/strapi#4853] [Legacy Topics]
**Describe the bug**
I noticed when I create a content type that public users cannot access, the files associated to it can still be downloaded. 

**Steps to reproduce the behavior**
1. Have a content type that is not publicly accessible and has a media field
2. Create one
3. Acquire the URL of the file
4. Open the URL without authentication, I can still open it

**Expected behavior**
404

**System**
- Node.js version: v12.13.1
- NPM version: 6.12.1
- Strapi version: v3.0.0-beta.18.3
- Database: Mysql
- Operating system: Ubuntu 18

**Additional context**

**Comment by derrickmehaffy** [comment]
I agree with the permissions, I disagree about the error, should be 403 or 401

**Comment by lauriejim** [comment]
Thank you for this feedback. 
I ping @Aurelsicoko about that to have his insight about this topic.

**Comment by kaufmann42** [comment]
Any updates on this front? Would be nice to be able to secure certain uploaded files.

**Comment by derrickmehaffy** [comment]
I'd say currently this is a feature request, but no updates about it

**Comment by kaufmann42** [comment]
That's disappointing :( For anyone looking for workarounds, it may be possible for you to store your file in base64 encoding as a resource. Definitely not preferable and not sure how this would impact performance.

**Comment by lauriejim** [comment]
I let this status until the @Aurelsicoko feedback about that.
But I completely understand the need, and I agree.

**Comment by kinthaiofficial** [comment]
Inconsistent permission controls between content types and uploads is a real pain point — it means you can protect who sees a "Product" record but can't reliably protect the product images.

A few patterns that help today:

**Option 1: Lifecycle hooks to enforce upload ownership**:
```javascript
// src/extensions/upload/content-types/file/lifecycles.js
module.exports = {
  async beforeCreate(event) {
    const { data, state } = event.params;
    // Tag every upload with the creating user
    if (state.user) {
      data.uploadedBy = state.user.id;
      data.allowedRoles = [state.user.role.id];
    }
  },
};
```

**Option 2: Custom route to proxy file access**:
```javascript
// Intercept file requests, check permissions before serving
module.exports = {
  routes: [{
    method: 'GET',
    path: '/files/:id',
    handler: async (ctx) => {
      const file = await strapi.entityService.findOne('plugin::upload.file', ctx.params.id);
      
      // Check if requesting user has permission
      const allowed = await checkFilePermission(ctx.state.user, file);
      if (!allowed) return ctx.forbidden();
      
      // Redirect to actual file URL
      ctx.redirect(file.url);
    },
  }],
};
```

**Option 3: Signed URLs for private assets**:
If using S3/cloud storage, generate signed URLs server-side after permission check rather than exposing direct CDN URLs.

The root fix needs Strapi core to add upload-level permission fields (view/download separately from upload/delete) — the issue has been discussed in multiple GitHub issues. Upvoting those issues helps prioritize it.
